homeassistant/components/sensor/__init__.py

```diff
@@ -302,36 +302,62 @@ def state(self) -> Any:
             # Anyways, lets validate the date at least..
             try:
                 value = ciso8601.parse_datetime(str(value))
-            except (ValueError, IndexError) as error:
-                raise ValueError(
-                    f"Invalid date/datetime: {self.entity_id} provide state '{value}', "
-                    f"while it has device class '{device_class}'"
-                ) from error
+            except (ValueError, IndexError):
+                _LOGGER.error(
+                    "Invalid date/datetime: %s provide state '%s', "
+                    "while it has device class '%s'",
+                    self.entity_id,
+                    value,
+                    device_class,
+                )
+                return None
 
             # Convert the date object to a standardized state string.
             if device_class == DEVICE_CLASS_DATE:
                 return value.date().isoformat()
+
+            if value.tzinfo is not None and value.tzinfo != timezone.utc:
+                value = value.astimezone(timezone.utc)
+
             return value.isoformat(timespec="seconds")
 
         # Received a datetime
         if value is not None and device_class == DEVICE_CLASS_TIMESTAMP:
             try:
-                return value.isoformat(timespec="seconds")  # type: ignore
-            except (AttributeError, TypeError) as err:
-                raise ValueError(
-                    f"Invalid datetime: {self.entity_id} has a timestamp device class"
-                    f"but does not provide a datetime state but {type(value)}"
-                ) from err
+                # We cast the value, to avoid using isinstance, but satisfy
+                # typechecking. The errors are guarded in this try.
+                value = cast(datetime, value)
+                if value.tzinfo is None:
+                    raise ValueError(
+                        f"Invalid datetime: {self.entity_id} provides state '{value}', "
+                        "which is missing timezone information"
+                    )
+
+                if value.tzinfo != timezone.utc:
+                    value = value.astimezone(timezone.utc)
+
+                return value.isoformat(timespec="seconds")
+            except (AttributeError, TypeError):
+                _LOGGER.error(
+                    "Invalid datetime: %s has a timestamp device class"
+                    "but does not provide a datetime state but %s",
+                    self.entity_id,
+                    type(value),
+                )
+                return None
```

Comment:
Why return None instead of raising? We risk masking this developer mistake.

Reply:
Because the result is more or less the same.

Example stack trace:

```
homeassistant.components.sensor:entity_platform.py:305 Error while setting up test platform for sensor
Traceback (most recent call last):
  File "/home/frenck/code/home-assistant/core/homeassistant/helpers/entity_platform.py", line 257, in _async_setup_platform
    await asyncio.gather(*pending)
  File "/home/frenck/code/home-assistant/core/homeassistant/helpers/entity_platform.py", line 382, in async_add_entities
    await asyncio.gather(*tasks)
  File "/home/frenck/code/home-assistant/core/homeassistant/helpers/entity_platform.py", line 607, in _async_add_entity
    await entity.add_to_platform_finish()
  File "/home/frenck/code/home-assistant/core/homeassistant/helpers/entity.py", line 717, in add_to_platform_finish
    self.async_write_ha_state()
  File "/home/frenck/code/home-assistant/core/homeassistant/helpers/entity.py", line 480, in async_write_ha_state
    self._async_write_ha_state()
  File "/home/frenck/code/home-assistant/core/homeassistant/helpers/entity.py", line 513, in _async_write_ha_state
    state = self._stringify_state()
  File "/home/frenck/code/home-assistant/core/homeassistant/helpers/entity.py", line 486, in _stringify_state
    if (state := self.state) is None:
  File "/home/frenck/code/home-assistant/core/tests/common.py", line 963, in state
    return self._handle("state")
  File "/home/frenck/code/home-assistant/core/tests/common.py", line 1004, in _handle
    return getattr(super(), attr)
  File "/home/frenck/code/home-assistant/core/homeassistant/components/sensor/__init__.py", line 325, in state
    raise ValueError(
ValueError: Invalid datetime: sensor.test_5 provides state '2017-12-19 18:29:42', which is missing timezone information
```

The ValueError provides the message, so does an error log.

That said, this doesn't happen in the path of the developer, as shown in the stack trace, the source of the problem isn't exposed (and thus this doesn't add value over the error log that is more user-friendly in case it does end up with the user).

Reply:
It shouldn't be a user friendly log. It should be clear that there's a bug and unexpected behavior.

Reply:
What would be the value of the stack trace here?

Please note, that these logs can originate from templates, MQTT, mobile apps and other external sources too, so this doesn't have to be "developer".

Reply:
Returning None is wrong. The state isn't unknown, we know that there's a bug. We should communicate this and stop any further processing of the state.

Reply:
That is actually a good point. Alright, reverting that part.

Reply:
Done!